common/ferrostar/src/routing_adapters/graphhopper.rs

```diff
@@ -0,0 +1,322 @@
+//! High-level HTTP request generation for GraphHopper-based HTTP APIs.
+
+use crate::models::{UserLocation, Waypoint};
+
+use crate::routing_adapters::error::{InstantiationError, RoutingRequestGenerationError};
+use crate::routing_adapters::{RouteRequest, RouteRequestGenerator};
+
+use serde::Serialize;
+use serde_json::{json, Map, Value as JsonValue};
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+#[derive(Serialize, Debug)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+#[serde(rename_all = "lowercase")]
+pub enum VoiceUnits {
+    Metric,
+    Imperial,
+}
+
+/// A route request generator for GraphHopper backends operating over HTTP.
+///
+/// # Examples
+///
+/// ```
+/// use serde_json::{json, Map, Value};
+/// use ferrostar::routing_adapters::graphhopper::{GraphHopperHttpRequestGenerator, VoiceUnits};
+/// let options: Map<String, Value> = json!({
+///     "ch.disable": true,
+///     "custom_model": {
+///         "distance_influence": 15,
+///         "speed": [
+///             {
+///                 "if": "road_class == MOTORWAY",
+///                 "limit_to": "100"
+///             }
+///         ]
+///     }
+/// }).as_object().unwrap().to_owned();
+/// let request_generator = GraphHopperHttpRequestGenerator::new(
+///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
+///     "car",
+///     "en",
+///     VoiceUnits::Metric, options
+/// );
+/// ```
+#[derive(Debug)]
+pub struct GraphHopperHttpRequestGenerator {
+    endpoint_url: String,
+    profile: String,
+    locale: String,
+    voice_units: VoiceUnits,
+    options: Map<String, JsonValue>,
+}
+
+impl GraphHopperHttpRequestGenerator {
+    /// Creates a new GraphHopper request generator given an endpoint URL, a profile name,
+    /// and options to include in the request JSON.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use serde_json::{json, Map, Value};
+    /// use ferrostar::routing_adapters::graphhopper::{GraphHopperHttpRequestGenerator, VoiceUnits};
+    /// let options: Map<String, Value> = json!({
+    ///     "ch.disable": true,
+    ///     "custom_model": {
+    ///         "distance_influence": 15,
+    ///         "speed": [
+    ///             {
+    ///                 "if": "road_class == MOTORWAY",
+    ///                 "limit_to": "100"
+    ///             }
+    ///         ]
+    ///     }
+    /// }).as_object().unwrap().to_owned();
+    /// let request_generator = GraphHopperHttpRequestGenerator::new(
+    ///     "https://graphhopper.com/api/1/navigate/?key=YOUR-API-KEY",
+    ///     "car",
+    ///     "en",
+    ///     VoiceUnits::Metric,
+    ///     options
+    /// );
+    /// ```
+    pub fn new<U: Into<String>, P: Into<String>, L: Into<String>>(
+        endpoint_url: U,
+        profile: P,
+        locale: L,
+        voice_units: VoiceUnits,
```

Comment:
With Valhalla, we leave locale and voice units unspecified, but you had them hard-coded, which makes me wonder if these are required and/or strongly encouraged for every request. I've made some changes to the interface assuming the answer is "yes" but please let me know your thoughts. For Valhalla, we currently just handle this with extra JSON options, but I'd like to move toward making the options easier to use with strong types, so here's a first pass at this.

cc @Archdoog in case you have any comments.